sktree/stats/forestht.py

```diff
@@ -0,0 +1,794 @@
+
+        Finally, resample the posteriors of the two forests to compute the null
+        distribution of the statistics.
+
+        Parameters
+        ----------
+        X : ArrayLike of shape (n_samples, n_features)
+            The data matrix.
+        y : ArrayLike of shape (n_samples, n_outputs)
+            The target matrix.
+        covariate_index : ArrayLike, optional of shape (n_covariates,)
+            The index array of covariates to shuffle, by default None.
+        metric : str, optional
+            The metric to compute, by default "mse".
+        n_repeats : int, optional
+            Number of times to sample the null distribution, by default 1000.
+        return_posteriors : bool, optional
+            Whether or not to return the posteriors, by default False.
+        **metric_kwargs : dict, optional
+            Additional keyword arguments to pass to the metric function.
+
+        Returns
+        -------
+        stat : float
+            The test statistic.
+        pval : float
+            The p-value of the test statistic.
+        """
+        X, y, covariate_index = self._check_input(X, y, covariate_index)
+
+        if self._n_samples_ is None:
+            # first compute the test statistic on the un-permuted data
+            observe_stat, observe_posteriors, observe_samples = self.statistic(
+                X,
+                y,
+                covariate_index=None,
+                metric=metric,
+                return_posteriors=True,
+                check_input=False,
+                **metric_kwargs,
+            )
+        else:
+            observe_samples = self.samples_
+            observe_posteriors = self.posterior_final_
+            observe_stat = self.stat_
+
+        # next permute the data
+        permute_stat, permute_posteriors, permute_samples = self.statistic(
+            X,
+            y,
+            covariate_index=covariate_index,
+            metric=metric,
+            return_posteriors=True,
+            check_input=False,
+            **metric_kwargs,
+        )
+
+        # Note: at this point, both `estimator` and `permuted_estimator_` should
+        # have been fitted already, so we can now compute on the null by resampling
+        # the posteriors and computing the test statistic on the resampled posteriors
+        if self.sample_dataset_per_tree:
+            metric_star, metric_star_pi = _compute_null_distribution_coleman(
+                y_test=y[observe_samples, :],
+                y_pred_proba_normal=observe_posteriors,
+                y_pred_proba_perm=permute_posteriors,
+                metric=metric,
+                n_repeats=n_repeats,
+                seed=self.random_state,
+            )
+        else:
+            if not self.sample_dataset_per_tree:
+                _, indices_test = self.train_test_samples_[0]
+                y_test = y[indices_test, :]
+            else:
+                y_test = y
```

Comment:
The first if statement should be `if not self.permute_per_tree`? The second condition now is redundant.